diff -up shadow-4.1.4.2/libmisc/user_busy.c.user-busy shadow-4.1.4.2/libmisc/user_busy.c
--- shadow-4.1.4.2/libmisc/user_busy.c.user-busy	2009-05-18 21:42:16.000000000 +0200
+++ shadow-4.1.4.2/libmisc/user_busy.c	2014-06-04 17:25:26.811346027 +0200
@@ -43,7 +43,7 @@
 
 #ifdef __linux__
 static int check_status (const char *sname, uid_t uid);
-static int user_busy_processes (uid_t uid);
+static int user_busy_processes (const char *name, uid_t uid);
 #else				/* !__linux__ */
 static int user_busy_utmp (const char *name);
 #endif				/* !__linux__ */
@@ -58,7 +58,7 @@ int user_busy (const char *name, uid_t u
 	 */
 #ifdef __linux__
 	/* On Linux, directly parse /proc */
-	return user_busy_processes (uid);
+	return user_busy_processes (name, uid);
 #else				/* !__linux__ */
 	/* If we cannot rely on /proc, check is there is a record in utmp
 	 * indicating that the user is still logged in */
@@ -91,6 +91,9 @@ static int user_busy_utmp (const char *n
 			continue;
 		}
 
+		fprintf (stderr,
+		         _("%s: user %s is currently logged in\n"),
+		         Prog, name);
 		return 1;
 	}
 
@@ -137,7 +140,7 @@ static int check_status (const char *sna
 	return 0;
 }
 
-static int user_busy_processes (uid_t uid)
+static int user_busy_processes (const char *name, uid_t uid)
 {
 	DIR *proc;
 	struct dirent *ent;
@@ -195,6 +198,9 @@ static int user_busy_processes (uid_t ui
 
 		if (check_status (tmp_d_name, uid) != 0) {
 			(void) closedir (proc);
+			fprintf (stderr,
+			         _("%s: user %s is currently used by process %d\n"),
+			         Prog, name, pid);
 			return 1;
 		}
 
@@ -212,6 +218,9 @@ static int user_busy_processes (uid_t ui
 				}
 				if (check_status (task_path+6, uid) != 0) {
 					(void) closedir (proc);
+					fprintf (stderr,
+					         _("%s: user %s is currently used by process %d\n"),
+					         Prog, name, pid);
 					return 1;
 				}
 			}
diff -up shadow-4.1.4.2/src/userdel.c.user-busy shadow-4.1.4.2/src/userdel.c
--- shadow-4.1.4.2/src/userdel.c.user-busy	2014-06-04 17:25:26.804345863 +0200
+++ shadow-4.1.4.2/src/userdel.c	2014-06-04 17:25:26.811346027 +0200
@@ -900,9 +900,6 @@ int main (int argc, char **argv)
 	 * a cron job may be started on her behalf, etc.
 	 */
 	if (user_busy (user_name, user_id) != 0) {
-		fprintf (stderr,
-		         _("%s: user %s is currently logged in\n"),
-		         Prog, user_name);
 		if (!fflg) {
 #ifdef WITH_AUDIT
 			audit_logger (AUDIT_DEL_USER, Prog,
diff -up shadow-4.1.4.2/src/usermod.c.user-busy shadow-4.1.4.2/src/usermod.c
--- shadow-4.1.4.2/src/usermod.c.user-busy	2014-06-04 17:25:26.000000000 +0200
+++ shadow-4.1.4.2/src/usermod.c	2014-06-04 17:47:37.439501239 +0200
@@ -1728,9 +1728,6 @@ int main (int argc, char **argv)
 	 */
 	if (   (uflg || lflg || dflg)
 	    && (user_busy (user_name, user_id) != 0)) {
-		fprintf (stderr,
-		         _("%s: user %s is currently logged in\n"),
-		         Prog, user_name);
 		exit (E_USER_BUSY);
 	}
 
